File: bukkit/elysium-chat-bukkit/src/main/java/com/seventh_root/elysium/chat/bukkit/chatchannel/pipeline/BukkitFormatChatChannelPipelineComponent.java
package com.seventh_root.elysium.chat.bukkit.chatchannel.pipeline;

import com.seventh_root.elysium.api.character.ElysiumCharacter;
import com.seventh_root.elysium.api.chat.ChatChannelPipelineComponent;
import com.seventh_root.elysium.api.chat.ChatMessageContext;
import com.seventh_root.elysium.api.chat.ElysiumChatChannel;
import com.seventh_root.elysium.api.player.ElysiumPlayer;
import com.seventh_root.elysium.characters.bukkit.character.BukkitCharacterProvider;
import com.seventh_root.elysium.chat.bukkit.ElysiumChatBukkit;
import com.seventh_root.elysium.chat.exception.ChatChannelMessageFormattingFailureException;
import com.seventh_root.elysium.core.bukkit.util.ChatColorUtils;
import org.bukkit.ChatColor;

import static com.seventh_root.elysium.api.chat.ChatChannelPipelineComponent.Type.FORMATTER;

public class BukkitFormatChatChannelPipelineComponent extends ChatChannelPipelineComponent {

    private final ElysiumChatBukkit plugin;
    private String formatString;

    public BukkitFormatChatChannelPipelineComponent(ElysiumChatBukkit plugin, String formatString) {
        this.plugin = plugin;
        this.formatString = formatString;
    }

    public void setFormatString(String formatString) {
        this.formatString = formatString;
    }

    @Override
    public Type getType() {
        return FORMATTER;
    }

    @Override
    public String process(String message, ChatMessageContext context) throws ChatChannelMessageFormattingFailureException {
        BukkitCharacterProvider characterProvider = plugin.getCore().getServiceManager().getServiceProvider(BukkitCharacterProvider.class);
        ElysiumPlayer sender = context.getSender();
        ElysiumPlayer receiver = context.getReceiver();
        ElysiumCharacter senderCharacter = characterProvider.getActiveCharacter(sender);
        ElysiumCharacter receiverCharacter = characterProvider.getActiveCharacter(receiver);
        ElysiumChatChannel chatChannel = context.getChatChannel();
        String formattedMessage = ChatColor.translateAlternateColorCodes('&', formatString);
        if (formattedMessage.contains("$message")) {
            if (message != null) {
                formattedMessage = formattedMessage.replace("$message", message);
            } else {
                throw new ChatChannelMessageFormattingFailureException(ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.no-message")));
            }
        }
        if (formattedMessage.contains("$sender-player")) {
            if (sender != null) {
                formattedMessage = formattedMessage.replace("$sender-player", sender.getName());
            } else {
                return null;
            }
        }
        if (formattedMessage.contains("$sender-character")) {
            if (senderCharacter != null) {
                formattedMessage = formattedMessage.replace("$sender-character", senderCharacter.getName());
            } else {
                throw new ChatChannelMessageFormattingFailureException(ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.no-character")));
            }
        }
        if (formattedMessage.contains("$receiver-player")) {
            if (receiver != null) {
                formattedMessage = formattedMessage.replace("$receiver-player", receiver.getName());
            } else {
                return null;
            }
        }
        if (formattedMessage.contains("$receiver-character")) {
            if (receiverCharacter != null) {
                formattedMessage = formattedMessage.replace("$receiver-character", receiverCharacter.getName());
            } else {
                return null;
            }
        }
        if (formattedMessage.contains("$channel")) {
            if (chatChannel != null) {
                formattedMessage = formattedMessage.replace("$channel", chatChannel.getName());
            } else {
                throw new ChatChannelMessageFormattingFailureException(ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.no-channel")));
            }
        }
        if (formattedMessage.contains("$color") || formattedMessage.contains("$colour")) {
            if (chatChannel != null) {
                String chatColorString = ChatColorUtils.closestChatColorToColor(chatChannel.getColor()).toString();
                formattedMessage = formattedMessage.replace("$color", chatColorString)
                        .replace("$colour", chatColorString);
            } else {
                throw new ChatChannelMessageFormattingFailureException(ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.no-channel")));
            }
        }
        return formattedMessage;
    }

}

File: bukkit/elysium-chat-bukkit/src/main/kotlin/com/seventh_root/elysium/chat/bukkit/chatchannel/pipeline/BukkitFormatChatChannelPipelineComponent.kt
package com.seventh_root.elysium.chat.bukkit.chatchannel.pipeline

import com.seventh_root.elysium.api.chat.ChatChannelPipelineComponent
import com.seventh_root.elysium.api.chat.ChatChannelPipelineComponent.Type.FORMATTER
import com.seventh_root.elysium.api.chat.ChatMessageContext
import com.seventh_root.elysium.api.chat.ChatMessagePostProcessContext
import com.seventh_root.elysium.api.chat.exception.ChatChannelMessageFormattingFailureException
import com.seventh_root.elysium.characters.bukkit.character.BukkitCharacterProvider
import com.seventh_root.elysium.chat.bukkit.ElysiumChatBukkit
import com.seventh_root.elysium.core.bukkit.util.ChatColorUtils
import org.bukkit.ChatColor

class BukkitFormatChatChannelPipelineComponent(private val plugin: ElysiumChatBukkit, var formatString: String?) : ChatChannelPipelineComponent() {

    override val type: ChatChannelPipelineComponent.Type
        get() = FORMATTER

    @Throws(ChatChannelMessageFormattingFailureException::class)
    override fun process(message: String, context: ChatMessageContext): String? {
        val characterProvider = plugin.core!!.serviceManager.getServiceProvider(BukkitCharacterProvider::class.java)
        val sender = context.sender
        val receiver = context.receiver
        val senderCharacter = characterProvider.getActiveCharacter(sender)
        val receiverCharacter = characterProvider.getActiveCharacter(receiver)
        val chatChannel = context.chatChannel
        var formattedMessage = ChatColor.translateAlternateColorCodes('&', formatString)
        if (formattedMessage.contains("\$message")) {
            formattedMessage = formattedMessage.replace("\$message", message)
        }
        if (formattedMessage.contains("\$sender-player")) {
            formattedMessage = formattedMessage.replace("\$sender-player", sender.name)
        }
        if (formattedMessage.contains("\$sender-character")) {
            if (senderCharacter != null) {
                formattedMessage = formattedMessage.replace("\$sender-character", senderCharacter.name)
            } else {
                throw ChatChannelMessageFormattingFailureException(ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.no-character")))
            }
        }
        if (formattedMessage.contains("\$receiver-player")) {
            formattedMessage = formattedMessage.replace("\$receiver-player", receiver.name)
        }
        if (formattedMessage.contains("\$receiver-character")) {
            if (receiverCharacter != null) {
                formattedMessage = formattedMessage.replace("\$receiver-character", receiverCharacter.name)
            } else {
                return null
            }
        }
        if (formattedMessage.contains("\$channel")) {
            formattedMessage = formattedMessage.replace("\$channel", chatChannel.name)
        }
        if (formattedMessage.contains("\$color") || formattedMessage.contains("\$colour")) {
            val chatColorString = ChatColorUtils.closestChatColorToColor(chatChannel.color).toString()
            formattedMessage = formattedMessage.replace("\$color", chatColorString).replace("\$colour", chatColorString)
        }
        return formattedMessage
    }

    override fun postProcess(message: String, context: ChatMessagePostProcessContext): String? {
        val characterProvider = plugin.core!!.serviceManager.getServiceProvider(BukkitCharacterProvider::class.java)
        val sender = context.sender
        val senderCharacter = characterProvider.getActiveCharacter(sender)
        val chatChannel = context.chatChannel
        var formattedMessage = ChatColor.translateAlternateColorCodes('&', formatString)
        if (formattedMessage.contains("\$message")) {
            formattedMessage = formattedMessage.replace("\$message", message)
        }
        if (formattedMessage.contains("\$sender-player")) {
            formattedMessage = formattedMessage.replace("\$sender-player", sender.name)
        }
        if (formattedMessage.contains("\$sender-character")) {
            if (senderCharacter != null) {
                formattedMessage = formattedMessage.replace("\$sender-character", senderCharacter.name)
            } else {
                throw ChatChannelMessageFormattingFailureException(ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.no-character")))
            }
        }
        if (formattedMessage.contains("\$receiver-player")) {
            formattedMessage = formattedMessage.replace("\$receiver-player", "\$receiver-player")
        }
        if (formattedMessage.contains("\$receiver-character")) {
            formattedMessage = formattedMessage.replace("\$receiver-character", "\$receiver-character")
        }
        if (formattedMessage.contains("\$channel")) {
            formattedMessage = formattedMessage.replace("\$channel", chatChannel.name)
        }
        if (formattedMessage.contains("\$color") || formattedMessage.contains("\$colour")) {
            val chatColorString = ChatColorUtils.closestChatColorToColor(chatChannel.color).toString()
            formattedMessage = formattedMessage.replace("\$color", chatColorString).replace("\$colour", chatColorString)
        }
        return formattedMessage
    }

}
